### ISSUE TITLE
Regular Expression for searches

### ISSUE THREAD

**Issue by notkaramel** [ECHOES-from-the-Past/mei-analyser#65]
- Following the wild cards examples from https://extras.humdrum.org/man/themax/
- To be implemented **_(this was discarded for the POSIX regex standard, see below)_**
  - [x] To search for one arbitrary pitch, represent it with a dot: `.`
  - [x] To search for an optional note, ~~append~~ **use** a question mark: `?`
    - Example: `d f? a` could match `d f a` or `d a`
    - Example: `d ? a` could match `d f a`, `d g a`, `d c a`, etc. and also`d a`.
  - [x] To search for any number of optional notes, insert an asterisk:`*` 
    - Example: `f * a` can match to `f a`, `f g a`, `f b g a`, etc. 

Based on these examples, this seems to be equivalent to the **non-greedy** version of the **`+` operator**: `.+?`. It stops at the first appearance of the end character (the non-greedy part given by the ending question mark `?`), and it can have no character between the start and end (as given by `.+`).

I propose we use the asterisk combined with `{,n}` to indicate the maximum number of characters between the start and end (as we know, the minimum is 0). For example: using `f *{,5} a` to indicate a sequence that starts with `f` and finishes with `a` that has no more than five notes in between. As long as we document its use, it is ok.

**Comment by notkaramel** [comment]
![image](https://github.com/user-attachments/assets/7c542493-d1b6-4e4a-94fe-841387481453)

**Comment by notkaramel** [comment]
@notkaramel Make a tooltip

**Comment by martha-thomae** [comment]
Most important things for next week:
1. Documentation on those four wildcards.
2. Highlight the results in the text-with-notes
3. Then you can continue implementing the wildcards.

Thank you!

**Comment by notkaramel** [comment]
@martha-thomae do you have any thoughts on the wording of the documentation/tooltip content shown below? I intentionally use "**use**" to mean the character can be used as a standalone, and "**append**" to mean the character can be used as a supporting character

- Use a dot `.` to search for one arbitrary pitch.
- Use or append a question mark `?` to search for an optional note.
  - Example: `d f? a` would match `d f a` or `d a`
  - Example: `d ? a` would match `d a`, `d f a`, `d g a`, `d c a`, etc.
- Use an asterisk `*` to search for any number of optional notes:
  - Example: `f * a` can match to `f a`, `f g a`, `f b g a`, etc.
- Append an at sign `@` to search for a repeated note:
  - Example: `f a@` would match `f a`, `f a a`, `f a a a`, etc.

**Comment by martha-thomae** [comment]
How about dividing it into two statements, each with its own examples? See the text below (I also changed the wording a bit to be more musicology-friendly):

* Use a dot `.` to search for one arbitrary pitch.
* Use a question mark `?` to search for an optional note. 
  * For example, searching for `d ? a` could return the following sequences of notes: `d a`, `d f a`, `d g a`, `d c a`, etc.
* Or use the question mark `?` after a particular pitch value to make that note optional. 
  * For example, searching for `d f? a` could return both the sequence `d a` and the sequence`d f a`.
* Use an asterisk `*` to search for any number of optional notes. 
  * For example, searching for `f * a` could return the following sequences: `f a`, `f g a`, `f b g a`, etc.
* Use an at sign `@` after a note to allow for repetitions of that note.
  * For example, searching for `f a@` could return the following sequences: `f a`, `f a a`, `f a a a`, etc.

**Comment by notkaramel** [comment]
@martha-thomae
- The logic for wildcard search is a lot different from the old pitch/loc matching. There are algorithms for those using recursive functions or dynamic programming, and so I am currently testing those out for the moment. [You can check out the wildcard pattern matching implementations here](https://www.geeksforgeeks.org/wildcard-pattern-matching/)
- The tooltip information looks a bit too wordy IMO. Maybe we can keep the instructions, but for the example we can do a table like the [**humdrum** example above](https://github.com/ECHOES-from-the-Past/mei-analyser/issues/65#issuecomment-2304800858)? 
> I should have given an update a while back, sorry!

**Comment by martha-thomae** [comment]
Answering your other comments:

> * The tooltip information looks a bit too wordy IMO. Maybe we can keep the instructions, but for the example we can do a table like the [**humdrum** example above](https://github.com/ECHOES-from-the-Past/mei-analyser/issues/65#issuecomment-2304800858)?

Yes, I agree with this. We will add this into the interface after the Sept. 25th meeting. Also with the implementation of the dynamic programming version for the wildcards.

**Comment by notkaramel** [comment]
So, instead of implementing search features via dynamic programming or iterative loops like things used to be (graph 1), I will implement a "construct regex serach" and match it against a string of pitches mapped from the NeumeComponent list (graph 2).

- Graph 1: ![image](https://github.com/user-attachments/assets/0fbf859a-c3ef-4895-b86f-3d94a84f155e)

- Graph 2: ![image](https://github.com/user-attachments/assets/395d6685-533f-4a3d-90d3-f6a056b92f61)

This method is way more easier to implement and to re-map back to the NeumeComponent list to obtain the XML:IDs for highlighting afterwards.

**Comment by notkaramel** [comment]
@martha-thomae With wildcard search coming to the interface (PR #88), I want to put it in the place of `Exact Pitch` melodic search option? The wildcard search is like the extended version of exact pitch search :)

**Comment by notkaramel** [comment]
@martha-thomae With wildcard search coming to the interface (PR #88), I want to put it in the place of `Exact Pitch` melodic search option. The wildcard search is like the extended version of exact pitch search :) What do you think?

**Comment by martha-thomae** [comment]
Yes, I agree! Thanks, Antoine!

**Comment by martha-thomae** [comment]
We can have search by pitch (where the user could use wildcards). And start by contour (where the user can give exact intervals or  general directions—i.e., up, down, same, neutral)

**Comment by notkaramel** [comment]
> Francesco:
> 
> > We need to check if we can set a maximum number of notes between two excerpts in the search. For example, if we have this expression `g * a`, we wouldn't like for the search to retrieve a full chant that starts in `g` and ends in `a`. We would like to be able to limit the amount of notes between `g` and `a`.
> 
> **Solution:** `g .{1,10} a` (no more than ten characters, but at least one). Here, the 'period' is repeated at most ten times and at least one time. I added a suggestion on how to implement this within our project in the[ issue description](https://github.com/ECHOES-from-the-Past/mei-analyser/issues/65#issue-2408010371). The solution is using the asterisk rather than the dot to be more consistent with other things implemented. As long as things are well documented in the interface, everything should be fine. We also can change the implementation in the future once we have feedback about it.

@martha-thomae I was reading through our conversation, turns out this issue that Francesco was asking has a solution using directly regexp `g .* a`.
![image](https://github.com/user-attachments/assets/316b95c0-8216-4b11-80f5-5982afbda1a7)

As for the **"limit the amount of notes between [two notes]"**, currently I implemented a filter for valid entry on the melodic pattern input box, so things like {1, 10} would be immediately dicarded. Will look into it/be discussed soon!

**Comment by notkaramel** [comment]
```regex
[.A-Ga-g]\{(\d+\,)?\d+\}
```
Would be the regexp to detect .{2,10} and variation of it (see image below)

![image](https://github.com/user-attachments/assets/2603c4e5-68af-442e-a736-52d6feafd5c9)

**Comment by notkaramel** [comment]
"Please implement @notkaramel! We like it"
